### PR TITLE
backfill sync from an anchor checkpoint state

### LINKED ISSUE #2556: BlocksByRange under WS
**Is your feature request related to a problem? Please describe.**

+ Implement this https://github.com/ethereum/eth2.0-specs/pull/2131/files
+ Part of #2530
+ Since it's quite isolated I make it a separate issue to avoid pulling us from being compliant to the whole spec

**Detail**
1. Since we only need to serve blocks within `MIN_EPOCHS_FOR_BLOCK_REQUESTS`  epochs, we can prune our db accordingly in order to avoid having our db grow forever
2. Validate `beacon_blocks_by_range` request accordingly
3. We have to backfill blocks if starting from a ws checkpoint
+ Don't need to get through state transition, just validate block proposer and parent-child relationship until recent block roots inside the ws state
+ Consider hooking into / add new sync type  https://github.com/ChainSafe/lodestar/blob/074649e35fe6a20a5aeab8898b67b73dca2e2900/packages/lodestar/src/sync/range/range.ts#L112

### PR DESCRIPTION
**Pull request by g11tech**
**Motivation**
This PR is to backfill in a scenario when wss checkpoint or any other anchor state is used to start the beacon node.

**Description**
Right now, this PR rebases https://github.com/ChainSafe/lodestar/pull/2637 on current master, making it workable base to build on top of the mentioned previous PR

Fixes:
- [x]  Don't reject the entire syncrange sequence if a block in mid is missing making it a non linear
- [x] Fixe the scenario where the syncrange keeps on resulting in a non anchored block at the end, i.e. end block is orphaned, and the previous head could be multi blocks behind, likely scenario in reorgs. Resolution, load the previous head in a solo request to set it as a new anchor
- [x] before rangeSync or sync from outside, check the DB as DB could have had blocks saved from previous run, no need to verify their signatures and blocks in db should already be verified.
- [x] Do a checkpoint check if a wss checkpoint has been provided which was in the past from the initialized beacon state (from DB because of a previous run). In case not provided compare against genesis when at genesis and panic if not same.
*Created a backfilledSequences repository module, which tracks backfilled sequences like 384:key=> 211:value i.e. backfilled  from slot 384 to 211 both including, in normal operation will typically store two sequences, one used by the backfill process and one used by the forward moving chain (which will upon finalizing will write a new sequence finalizedSlot=>chainStarted/anchorSlot). Any gaps in the db because of restarting with checkpoint sync will result into a new sequence entry, but the previous entry will be eliminated when the backfill syncs up the gap*
- [x] Strategy to skip already filled sequence (needs a strategy to track gaps or filled sequences)
- [x]  `InvalidWSCheckpoint` error (https://github.com/ChainSafe/lodestar/issues/3385)
- [x] Schedule the processing/verification when the chain is supposedly free *throttling chain by `BATCH_SIZE` in a single trigger cycle (in db reads or syncRange fetches)*
- [x] grafana dashboard

~~skipBackFillSync option?
backfillTill:  options i) till  wsCheckpoint  to be validated, ii)till GENESIS_SLOT (default) iii)till provided Slot~~

Tested Scenarios:
- [x] sync test case: sync/wss.test.ts
image
- [x] prater backfill sync to genesis (despite multiple restarts interrupting the sync, recovering using backfilled Sequences)
image
- [x] stopping and starting multiple times creating gaps in the back filled sequences, backfiled sync filled the gaps and combined them with backfilled sequences to avoid redoing previous backfill sync. 
*example of backfillSync  recovering on a restart using previous backfilled sequence (jumped back from `1633696` to `1437564`, and then continuing backfilling from there on*:
image
- [x] recovering from a persistent non anchored sequence in range sync (because of the orphaned block being served at the edge of range sync)
image
- [x] recovering the **backfilled to genesis** state on a fully backfilled node restart
image
- [x] stopped a fully backfilled node and restarted after a gap with new recent checkpoint - backfilled from the new recent checkpoint to previous finalized checkpoint when the node was stopped, and then correlated with already backfilled sequence till the GENESIS to switch to backfilled complete state
image

Closes #2556

### CODE REVIEW

**Comment by g11tech** [comment]
~~example of backfillSync  recovering on a restart using previous backfilled sequence (jumped back from `1633696` to `1437564`, and then continuing backfilling from there on~~ (updated the scenario in the Tested Scenario section)

File: packages/lodestar/src/metrics/metrics/lodestar.ts [hunk 1]
```diff
@@ -411,6 +411,21 @@ export function createLodestarMetrics(
       }),
     },
 
+    backfillSync: {
+      totalBlocks: register.gauge({
+        name: "lodestar__backfill_sync_total_blocks",
```
Comment (dapplion) [inline packages/lodestar/src/metrics/metrics/lodestar.ts]: Remove double underscore from `lodestar__backfill`

File: packages/lodestar/src/node/notifier.ts [hunk 2]
```diff
@@ -59,6 +59,11 @@ export async function runNodeNotifier({
       const skippedSlots = clockSlot - headInfo.slot;
       const clockSlotRow = `slot: ${clockSlot}` + (skippedSlots > 0 ? ` (skipped ${skippedSlots})` : "");
 
+      // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
```
Comment (dapplion) [inline packages/lodestar/src/node/notifier.ts]: Don't disable this rule, use strict comparisions even if it's more verbose

File: packages/lodestar/src/node/notifier.ts [hunk 3]
```diff
@@ -75,20 +80,21 @@ export async function runNodeNotifier({
             clockSlotRow,
             headRow,
             finalizedCheckpointRow,
+            ...backfilledStat,
```
Comment (dapplion) [inline packages/lodestar/src/node/notifier.ts]: I won't add any info to the notifier. Users don't really care about backfill status. We must ensure it works and they trust us that it does. We should use metrics and API debug endpoints to learn about the status of backfill sync

File: packages/lodestar/src/sync/backfill/backfill.ts [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+  private readonly chain: IBeaconChain;
+  private readonly network: INetwork;
+  private readonly db: IBeaconDb;
+  private readonly config: IBeaconConfig;
+  private readonly logger: ILogger;
+  private readonly metrics: IMetrics | null;
+  private wsCheckpoint?: phase0.Checkpoint | null;
+  private opts: BackfillSyncOpts;
+
+  // At any given point, we should have either the anchorBlock, or anchorBlockRoot, the reversed head of the backfill sync, from where we need to backfill
```
Comment (dapplion) [inline packages/lodestar/src/sync/backfill/backfill.ts]: Limit line width to strictly less than 120 characters wide for readability

File: packages/lodestar/src/sync/backfill/backfill.ts [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        this.backfillStartFromSlot = anchorCp.epoch * SLOTS_PER_EPOCH;
+        this.logger.info("BackfillSync - initializing from Checkpoint", {
+          root: toHexString(anchorCp.root),
+          epoch: anchorCp.epoch,
+        });
+      }
+
+      // Update from the backfilled sequence
+
+      // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
```
Comment (dapplion) [inline packages/lodestar/src/sync/backfill/backfill.ts]: Please remove all disablings of the rule `@typescript-eslint/strict-boolean-expressions` from this PR

File: packages/lodestar/src/sync/backfill/backfill.ts [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+  /** Throw / return all AsyncGenerators */
+  close(): void {
+    this.network.events.off(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.off(NetworkEvent.peerDisconnected, this.removePeer);
+    this.processor.end(new ErrorAborted("BackfillSync"));
+  }
+
+  /** Returns oldestSlotSynced */
+  private async sync(): Promise<Slot> {
```
Comment (dapplion) [inline packages/lodestar/src/sync/backfill/backfill.ts:280]: This function is 134 lines long and hard to follow, please
- Add code comments throughout the code blocks explaining what is going on and how this is intended to work
- If possible try to modularize the code with good abstractions that help the above point self-documenting the code
- As bonus (if applicable) you can add a diagram or summary of the entire flow at the top of the function

File: packages/lodestar/src/sync/backfill/errors.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import {LodestarError} from "@chainsafe/lodestar-utils";
+import PeerId from "peer-id";
+import {Root} from "@chainsafe/lodestar-types";
+
+export enum BackfillSyncErrorCode {
+  /** fetched block doesn't connect to anchor block */
+  NOT_ANCHORED = "not_anchored",
+  // fetched blocks are not linear
```
Comment (dapplion) [inline packages/lodestar/src/sync/backfill/errors.ts]: Make all these comments JSDocs
```ts
/** fetched blocks are not linear */
```

File: packages/lodestar/src/sync/backfill/verify.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+): {
+  nextAnchor: allForkTypes.SignedBeaconBlock | null;
+  verifiedBlocks: allForkTypes.SignedBeaconBlock[];
+  error?: BackfillSyncErrorCode.NOT_LINEAR;
+} {
+  let nextRoot: Root = anchorRoot;
+  let nextAnchor: allForkTypes.SignedBeaconBlock | null = null;
+
+  const verifiedBlocks: allForkTypes.SignedBeaconBlock[] = [];
+  for (const block of blocks.slice(0).reverse()) {
```
Comment (dapplion) [inline packages/lodestar/src/sync/backfill/verify.ts]: Just loop backwards (with a regular for loop) without allocating a new array twice

File: packages/lodestar/src/sync/backfill/verify.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+export async function verifyBlockProposerSignature(
+  bls: IBlsVerifier,
+  state: CachedBeaconState<allForks.BeaconState>,
+  blocks: allForkTypes.SignedBeaconBlock[]
+): Promise<void> {
+  const signatures = blocks
+    // genesis block doesn't have valid signature
+    .filter((block) => block.message.slot !== GENESIS_SLOT)
+    .map((block) => allForks.getProposerSignatureSet(state, block));
```
Comment (dapplion) [inline packages/lodestar/src/sync/backfill/verify.ts]: .filter and .map can be merged into a single loop to prevent looping twice

File: packages/lodestar/src/sync/sync.ts [hunk 10]
```diff
@@ -21,6 +24,9 @@ export class BeaconSync implements IBeaconSync {
   private readonly rangeSync: RangeSync;
   private readonly unknownBlockSync: UnknownBlockSync;
 
+  // avoid finding same root at the same time
+  private readonly processingRoots = new Set<string>();
```
Comment (dapplion) [inline packages/lodestar/src/sync/sync.ts]: Is this used?

Comment (g11tech) [reply packages/lodestar/src/sync/sync.ts]: nope, will remove this :+1:

**Comment by dapplion** [comment]
@g11tech Remember to mark as "Ready for review" when appropiate! Draft PRs should be a transient state for unfinished work

**Comment by g11tech** [comment]
Graphana dashboard:
image
analysis: 
1. on starting beacon node with weakSubjectivitySync latest, the backfill sync started (1st graph), 
2. backfillTillSlot (2nd graph) started decreasing via syncing blocks from the network using rangesync (blue line in 3rd graph), 
3. till around 17:50 time mark, where backfill sync encountered a previous backfilled saved sequence (from the previous run few days ago) all the way to genesis (yellow spike in 3rd graph).
4. then backfill sync status turns to completed, backfillTillSlot goes to 0 and backfill sync exits.

4rth graph is the error rate.

**Comment by dapplion** [comment]
This looks super good! Thank you so much for the detailed charts and progression :heart: Will do another review in depth asap and should be good to go.

**Comment by g11tech** [comment]
> This looks super good! Thank you so much for the detailed charts and progression heart Will do another review in depth asap and should be good to go.

thank you! :+1:

File: packages/lodestar/src/chain/interface.ts [hunk 11]
```diff
@@ -40,6 +40,7 @@ export interface IBeaconChain {
   readonly executionEngine: IExecutionEngine;
   // Expose config for convenience in modularized functions
   readonly config: IBeaconConfig;
+  readonly anchorSlot: Slot;
```
Comment (wemeetagain) [inline packages/lodestar/src/chain/interface.ts:44]: would be nice to have a js doc:

Something like:
The initial slot that the chain is started with.

File: packages/lodestar/src/db/beacon.ts [hunk 12]
```diff
@@ -53,6 +54,8 @@ export class BeaconDb extends DatabaseService implements IBeaconDb {
   lightClientInitProofIndex: LightClientInitProofIndexRepository;
   lightClientSyncCommitteeProof: LightClientSyncCommitteeProofRepository;
 
+  backfilledSequences: BackfilledSequences;
```
Comment (wemeetagain) [inline packages/lodestar/src/db/beacon.ts]: I think using the term "ranges" would be more descriptive than "sequences".

**Review by wemeetagain — CHANGES_REQUESTED** [review]

File: packages/cli/src/cmds/beacon/initBeaconState.ts [hunk 13]
```diff
@@ -134,16 +135,19 @@ export async function initBeaconState(
...
+    const anchorState = await initStateFromAnchorState(config, db, logger, lastDbState);
+    return {anchorState};
   } else {
     const genesisStateFile = args.genesisStateFile || getGenesisFileUrl(args.network || defaultNetwork);
     if (genesisStateFile && !args.forceGenesis) {
       const stateBytes = await downloadOrLoadFile(genesisStateFile);
       const anchorState = getStateTypeFromBytes(chainForkConfig, stateBytes).createTreeBackedFromBytes(stateBytes);
       const config = createIBeaconConfig(chainForkConfig, anchorState.genesisValidatorsRoot);
-      return await initStateFromAnchorState(config, db, logger, anchorState);
+      await initStateFromAnchorState(config, db, logger, anchorState);
```
Comment (wemeetagain) [inline packages/cli/src/cmds/beacon/initBeaconState.ts]: I think we should be returning the return value of initStateFromAnchorState here

**Review by wemeetagain — DISMISSED** [review]

File: packages/cli/src/cmds/beacon/initBeaconState.ts (hunk 13 above)
Comment (g11tech) [reply packages/cli/src/cmds/beacon/initBeaconState.ts]: `initStateFromAnchorState` returns anchorState itself, but yes it would look cleaner

**Comment by g11tech** [comment]
synopsis of a test run with the latest PR code:

image
1. Started backfill sync around 16:30
2. Restarted coouple of times between ~17:00-18:30, sync recovered the previous progress using backfillRanges
3. Synced to genesis at 20:00 but sync aborted with failiure as code was checking signature of genesis block as well
4. Fixed it around 22:00, and the sync recovered previous progress to genesis which it complete successfully

@dapplion : Kindly review and let know if anything further needs to be fine tuned.

**Comment by g11tech** [comment]
@dapplion updated the PR with the discussed changes (and rebased with current master). test run where backfill sync recovered previous progress and completed sync to genesis. :slightly_smiling_face: 

image

**Review by dapplion — APPROVED** [review]
Excellent! Looking forward to many more like this in 2022 :tada:

**Comment by g11tech** [comment]
> Excellent! Looking forward to many more like this in 2022 tada

Thank you! lodestar rocks! :rocket: 
latest run on prater
image
